File: Java/Civitas/src/civitas/MazoSorpresas.java
package civitas;
import java.util.ArrayList;
import java.util.Collections;

public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init(){
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    public MazoSorpresas(){
        init();
        debug = false;
    }
    
    public MazoSorpresas(boolean tdebug){
        init();
        debug = tdebug;
        Diario.getInstance().ocurreEvento("MazoSorpresas -- modo debug : " + debug);
    }
    
    public void alMazo(Sorpresa s){
        if(!barajada){
            sorpresas.add(s);
        }
    }
    
    public Sorpresa siguiente(){
        if(!barajada || usadas >= sorpresas.size()){
            Collections.shuffle(sorpresas);
            usadas = 0;
            barajada = true;
        }
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.set(0, sorpresas.get(sorpresas.size()-1));
        sorpresas.set(sorpresas.size()-1, ultimaSorpresa);
        usadas++;
        return ultimaSorpresa;
    }
    
    public void inhabilitarCartaEspecial(Sorpresa s){
        int i = sorpresas.indexOf(s);
        if(i >= 0){
            sorpresas.remove(i);
            cartasEspeciales.add(s);
            Diario.getInstance().ocurreEvento("MazoSorpresas -- "
                    + "Carta especial deshabilitada");
        }
    }
    
    public void habilitarCartaEspecial(Sorpresa s){
        int i = cartasEspeciales.indexOf(s);
        if(i >= 0){
            cartasEspeciales.remove(i);
            sorpresas.add(s);
            Diario.getInstance().ocurreEvento("MazoSorpresas -- "
                    + "Carta especial habilitada");
        }
    }
    
}

File: Ruby/CivitasR/lib/MazoSorpresas.rb
#encoding:utf-8
require_relative "Sorpresa"
require_relative "Diario"

module Civitas
  class MazoSorpresas
    def init
      @barajada=false
      @usadas=0
      @sorpresas=[]
      @cartasEspeciales=[]
      @ultimaSorpresa
    end


    def initialize(n=nil)
      if n != nil
        @debug= n
        Diario.instance.ocurre_evento("Diario -- Se inicializa el mazo con debug = " + n.to_s)
      else
        @debug = false
      end
      init
    end

    def alMazo(s)
      @barajada ? nil : @sorpresas.push(s)
    end

    def siguiente
      if(!@barajadas || @usadas==@sorpresas.length || !@debug)
        @sorpresas.shuffle
        @usadas=0
        @barajada=true
      end
      @usadas+=1
      copia=@sorpresas[0]
      @sorpresas.shift
      @sorpresas.push(copia)
      @ultimaSorpresa=copia
      return @ultimaSorpresa
    end

    def inhabilitarCartaEspecial(s)   
      p=false
      for g in @sorpresas do
        if s==g
          p=true
          @sorpresas.delete(g)
        end
      end
      if(p)
        @cartasEspeciales.push(s)
      end
      Diario.instance.ocurre_evento("Se ha eliminado la carta especial "+s.to_s)
    end

    def habilitarCartaEspecial(s)
      p=false
      for g in @cartasEspeciales do
        if(g==s)
          p=true
          @cartasEspeciales.delete(g)
        end
      end
      if(p)
        @sorpresas.push(s)
      end
      Diario.instance.ocurre_evento("Se ha añadido la carta especial "+s.to_s)
    end
    private :init
  end
end